### PR TITLE
Suppress nullable conversion warnings in default arguments

### LINKED ISSUE #51622: Spurious CS8620 in SDK 5.0.200 (post VS 16.9 update) for nullable enum parameters with non-null default values.
**Version Used**: 
SDK Version: 5.0.200
**Steps to Reproduce**:

Program.cs:
```C#
#nullable enable
using System;
Bla();    //CS8620: Argument of type 'int' cannot be used for parameter 'x' of
            //type 'DayOfWeek?' in 'void Bla(DayOfWeek? x = 1)' due to
            //differences in the nullability of reference types.

static void Bla(DayOfWeek? x = DayOfWeek.Monday) { }
```
project file:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>
</Project>
```

**Expected Behavior**:
Compiles and runs without warning.

**Actual Behavior**:

warning CS8620: Argument of type 'int' cannot be used for parameter 'x' of type 'DayOfWeek?' in 'void Bla(DayOfWeek? x = 1)' due to differences in the nullability of reference types.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #51622

The issue is basically that the binder creates a bound node for the argument with a conversion whose "isCast" is false. It seems like the nullable walker will try to recreate such conversions with nullable types re-inferred to try and decide if they're valid. We try to re-classify the conversion with the assumption that the conversion must be implicit, since there seemingly is no cast.

In effect, `MyEnum? e = MyEnum.Value` currently works properly when `MyEnum.Value` has constant value 0, but but not when the constant value is non-zero. This is because the default argument that gets inserted is of the underlying type converted to the parameter type--`MyEnum? e = 0` is allowed, but `MyEnum? e = 1` is not. You have to use `MyEnum? e = (MyEnum?)1`

It's also not enough to just always set isCast: true--this ends up breaking certain tests where flow analysis depends on the values of default arguments due to attributes. So I decided to set it based on whether the default argument conversion that we found is an explicit conversion.

Ultimately, nullable diagnostics are never supposed to be reported on implicit default arguments, including the conversions of such arguments, and that's the only change in this PR that's strictly needed to resolve this bug. However, it also seems reasonable to adjust the creation of default arguments for consistency in the public API.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -35283,6 +35283,30 @@ public struct MyStruct
                 Diagnostic(ErrorCode.ERR_NoConversionForNubDefaultParam, "s").WithArguments("MyStruct", "s").WithLocation(4, 30));
         }
 
+        [Fact]
+        [WorkItem(51622, "https://github.com/dotnet/roslyn/issues/51622")]
+        public void NullableDefaultArgument_NullableEnumFromMetadata()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:35288]:
```suggestion
        public void NullableEnumDefaultArgument_NonZeroValue()
```

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IArgument.cs [hunk 2]
```diff
@@ -4035,6 +4036,54 @@ public struct MyStruct
             comp.VerifyEmitDiagnostics(expectedDiagnostics);
         }
 
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        [WorkItem(39868, "https://github.com/dotnet/roslyn/issues/39868")]
+        public void NullableDefaultArgument_NullableEnumFromMetadata()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IArgument.cs:4042]:
```suggestion
        public void NullableEnumDefaultArgument_NonZeroValue()
```

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 3]
```diff
@@ -1411,7 +1411,14 @@ BoundExpression bindDefaultArgument(SyntaxNode syntax, ParameterSymbol parameter
                     {
                         GenerateImplicitConversionError(diagnostics, syntax, conversion, defaultValue, parameterType);
                     }
-                    defaultValue = CreateConversion(defaultValue, conversion, parameterType, diagnostics);
+                    defaultValue = CreateConversion(
+                        defaultValue.Syntax,
+                        defaultValue,
+                        conversion,
+                        isCast: conversion.IsExplicit,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1418]: From our offline discussion: is it possible to observe this change?

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 4]
```diff
@@ -1411,7 +1411,14 @@ BoundExpression bindDefaultArgument(SyntaxNode syntax, ParameterSymbol parameter
...
+                        new ConversionGroup(conversion, parameter.TypeWithAnnotations),
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1419]: From our offline discussion: consider avoiding this allocation when `isCast` is `false`

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 5)

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1418]: It does not appear to be observable through either public API or through differences in diagnostics.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1418]: Oddly enough I think the IsImplicit property on the IConversionOperation ends up being true if either the node is compiler generated, or if the conversion is not from a cast

https://github.com/dotnet/roslyn/blob/7e8b9ee7491b4d51809f5cc0273883692807c279/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs#L952

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1418]: I made a pass over the usages of 'BoundConversion.IsExplicitCastInCode' and I wasn't able to come up with a scenario where behavior for a default argument would change.

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1419]: Done

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6)

**Pull request merged**